tests/dragonfly/utility.py

```diff
@@ -15,68 +20,413 @@ def grouper(n, iterable):
-    for group in grouper(BATCH_SIZE, gen):
-        vals = [as_str_val(v) for v in client.mget(k for k, _ in group)]
-        gvals = [v for _, v in group]
-        assert vals == gvals
+    def __init__(self, target_keys, val_size, batch_size, max_multikey):
+        self.key_cnt_target = target_keys
+        self.val_size = val_size
+        self.batch_size = min(batch_size, target_keys)
+        self.max_multikey = max_multikey
 
-async def batch_check_data_async(client: aioredis.Redis, gen):
-    for group in grouper(BATCH_SIZE, gen):
-        vals = [as_str_val(v) for v in await client.mget(k for k, _ in group)]
-        gvals = [v for _, v in group]
-        assert vals == gvals
+        # Key management
+        self.key_sets = [set() for _ in ValueType]
+        self.key_cursor = 0
+        self.key_cnt = 0
 
-def wait_available(client: redis.Redis):
-    its = 0
-    while True:
-        try:
-            client.get('key')
-            print("wait_available iterations:", its)
-            return
-        except redis.ResponseError as e:
-            assert "Can not execute during LOADING" in str(e)
+        # Grow factors
+        self.diff_speed = 5
+        self.base_diff_prob = 0.2
+        self.min_diff_prob = 0.1
 
-        time.sleep(0.01)
-        its += 1
+    def keys(self):
+        return itertools.chain(*self.key_sets)
 
+    def keys_and_types(self):
+        return ((k, t) for t in list(ValueType) for k in self.set_for_type(t))
 
-async def wait_available_async(client: aioredis.Redis):
-    its = 0
-    while True:
-        try:
-            await client.get('key')
-            print("wait_available iterations:", its)
-            return
-        except aioredis.ResponseError as e:
-            assert "Can not execute during LOADING" in str(e)
+    def set_for_type(self, t: ValueType):
+        return self.key_sets[t.value]
 
-        await asyncio.sleep(0.01)
-        its += 1
+    def add_key(self, t: ValueType):
+        k, self.key_cursor = self.key_cursor, self.key_cursor + 1
+        self.set_for_type(t).add(k)
+        return k
+
+    def randomize_type(self):
+        return random.choice([t for t in ValueType])
+
+    def randomize_nonempty_set(self):
+        if not any(self.key_sets):
+            return None, None
+
+        t = self.randomize_type()
+        s = self.set_for_type(t)
+
+        if len(s) == 0:
+            return self.randomize_nonempty_set()
+        else:
+            return s, t
+
+    def randomize_key(self, t=None, pop=False):
+        if t is None:
+            s, t = self.randomize_nonempty_set()
+        else:
+            s = self.set_for_type(t)
+
+        if s is None or len(s) == 0:
+            return None, None
+
+        k = s.pop()
+        if not pop:
+            s.add(k)
+
+        return k, t
+
+    def generate_val(self, t: ValueType):
+        def rand_str(k=3, s=''):
+            return s.join(random.choices(string.ascii_letters, k=k))
+
+        if t == ValueType.STRING:
```

Comment:
maybe to much time with Rust, but I think that pattern matching in this case it better https://www.infoworld.com/article/3609208/how-to-use-structural-pattern-matching-in-python.html

Reply:
I know it exists, but its Python 3.10, whereas most repos of most distros cover for now only Python 3.9

I would rather not use fancy new features so that its runnable with default python and doesn't require installing a new version just to do it